File: app/src/lib/ig/cripto.ts
import "server-only";

import {
  createCipheriv,
  createDecipheriv,
  randomBytes,
  type CipherGCMTypes,
} from "node:crypto";

import { requireServerEnv } from "@/lib/env/server";

/**
 * O token do Instagram em repouso (docs/PLANO.md, "Segurança" §3).
 *
 * AES-256-GCM, chave em `TOKEN_ENC_KEY` (32 bytes em base64, validada em
 * `lib/env/server.ts`), **IV novo a cada gravação**. GCM com IV repetido não
 * degrada um pouco: repetir o par (chave, IV) entrega o keystream e, pior,
 * permite forjar autenticação para aquela chave. Por isso o IV nunca vem de
 * contador nem de hash do conteúdo — só de `randomBytes`.
 *
 * O DADO ASSOCIADO (AAD) É O `ig_user_id`
 * =======================================
 *
 * GCM autentica o texto cifrado; o AAD estende essa autenticação a um dado que
 * NÃO é cifrado. Aqui ele amarra o token à conta dona dele.
 *
 * Sem AAD, alguém com escrita no banco (e sem a chave) poderia copiar
 * `token_cipher`/`token_iv`/`token_tag` da linha da conta A para a linha da
 * conta B. A decifragem daria certo — é o mesmo texto, a mesma chave — e o
 * servidor publicaria no perfil A achando que era o B. Com o `ig_user_id` como
 * AAD, a troca faz a verificação da tag falhar e `decifrar` lança.
 *
 * Consequência prática: quem cifra e quem decifra precisam passar o MESMO
 * `ig_user_id`. É por isso que `ig_accounts_para_renovar` (migration 0018) traz
 * essa coluna junto com o token.
 */

const ALGORITMO: CipherGCMTypes = "aes-256-gcm";

/** 96 bits é o tamanho para o qual o GCM foi especificado e otimizado. */
const BYTES_DO_IV = 12;

/**
 * A versão da chave que este código usa para CIFRAR. A coluna `key_version` em
 * `ig_accounts` guarda com qual versão cada linha foi cifrada, para uma rotação
 * futura poder decifrar o que ficou para trás sem downtime (RUNBOOK).
 */
export const VERSAO_DA_CHAVE = 1;

export type TokenCifrado = {
  cipherHex: string;
  ivHex: string;
  tagHex: string;
  keyVersion: number;
};

function chave(): Buffer {
  // `requireServerEnv` já garante presença; o schema do zod já garante os 32
  // bytes. Aqui só se decodifica.
  return Buffer.from(requireServerEnv("TOKEN_ENC_KEY"), "base64");
}

export function cifrar(texto: string, igUserId: string): TokenCifrado {
  const iv = randomBytes(BYTES_DO_IV);
  const cifra = createCipheriv(ALGORITMO, chave(), iv);
  cifra.setAAD(Buffer.from(igUserId, "utf8"));

  const cipher = Buffer.concat([
    cifra.update(texto, "utf8"),
    cifra.final(),
  ]);

  return {
    cipherHex: cipher.toString("hex"),
    ivHex: iv.toString("hex"),
    tagHex: cifra.getAuthTag().toString("hex"),
    keyVersion: VERSAO_DA_CHAVE,
  };
}

/**
 * Lança quando a tag não confere — ou seja, quando o texto cifrado foi
 * adulterado, quando a chave mudou, ou quando o `ig_user_id` não é o da linha.
 *
 * A mensagem **não** carrega nada do conteúdo. Um erro de decifragem que
 * imprimisse bytes do token seria exatamente o vazamento que o resto do arquivo
 * existe para evitar.
 */
export function decifrar(guardado: TokenCifrado, igUserId: string): string {
  try {
    const decifra = createDecipheriv(
      ALGORITMO,
      chave(),
      Buffer.from(guardado.ivHex, "hex"),
    );
    decifra.setAAD(Buffer.from(igUserId, "utf8"));
    decifra.setAuthTag(Buffer.from(guardado.tagHex, "hex"));

    return Buffer.concat([
      decifra.update(Buffer.from(guardado.cipherHex, "hex")),
      decifra.final(),
    ]).toString("utf8");
  } catch {
    throw new Error(
      "nao foi possivel decifrar o token desta conta " +
        `(key_version=${guardado.keyVersion})`,
    );
  }
}

File: worker/src/servico/cripto.py
"""Decifra o token do Instagram em repouso — o espelho de `app/src/lib/ig/cripto.ts`.

AES-256-GCM, chave de 32 bytes em base64 (`TOKEN_ENC_KEY`), IV de 12 bytes por
linha e o `ig_user_id` como dado associado (AAD). Os dois lados precisam
concordar em cada detalhe, e por isso este arquivo e curto e sem opcao: nao ha
parametro para escolher algoritmo, tamanho de IV nem AAD.

O AAD e o que amarra o token a conta dona dele. Um texto cifrado copiado para a
linha de outra conta nao decifra — a tag nao confere — e `TokenIndecifravel` e
tudo que sai daqui. A mensagem nunca carrega byte nenhum do conteudo.

O worker so DECIFRA. Cifrar e do app (callback do OAuth e renovacao); o worker
nunca grava token nenhum.
"""
from __future__ import annotations

import base64
import binascii

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

TAMANHO_DA_CHAVE = 32
TAMANHO_DO_IV = 12


class ChaveInvalida(ValueError):
    pass


class TokenIndecifravel(RuntimeError):
    """A tag nao conferiu, a chave e outra, ou o `ig_user_id` nao e o da linha."""


def carregar_chave(base64_texto: str) -> bytes:
    try:
        chave = base64.b64decode(base64_texto, validate=True)
    except (binascii.Error, ValueError) as erro:
        raise ChaveInvalida("TOKEN_ENC_KEY nao e base64 valido.") from erro
    if len(chave) != TAMANHO_DA_CHAVE:
        raise ChaveInvalida(
            f"TOKEN_ENC_KEY precisa ter {TAMANHO_DA_CHAVE} bytes (openssl rand -base64 32)."
        )
    return chave


def decifrar(chave: bytes, cipher_hex: str, iv_hex: str, tag_hex: str, aad: str) -> str:
    try:
        iv = bytes.fromhex(iv_hex)
        cifrado = bytes.fromhex(cipher_hex)
        tag = bytes.fromhex(tag_hex)
    except (ValueError, TypeError) as erro:
        raise TokenIndecifravel("token cifrado fora do formato esperado") from erro

    if len(iv) != TAMANHO_DO_IV:
        raise TokenIndecifravel("IV fora do tamanho esperado")

    try:
        # A `cryptography` espera cifra+tag concatenados; o Node os separa.
        aberto = AESGCM(chave).decrypt(iv, cifrado + tag, aad.encode("utf-8"))
    except InvalidTag as erro:
        raise TokenIndecifravel("a tag de autenticacao nao confere") from erro

    try:
        return aberto.decode("utf-8")
    except UnicodeDecodeError as erro:
        raise TokenIndecifravel("token decifrado nao e texto") from erro
